tests/cpgrid/LgrChecks.hpp

```diff
@@ -0,0 +1,466 @@
+
+    for (const auto& element : elements(leafView)) {
+        BOOST_CHECK_EQUAL(element.index(), leafMapper.index(element));
+    }
+
+    const int maxLevel = grid.maxLevel();
+    for (int level = 0; level <= maxLevel; ++level) {
+        const auto& levelView = grid.levelGridView(level);
+        Dune::MultipleCodimMultipleGeomTypeMapper levelMapper(levelView, Dune::mcmgElementLayout());
+
+        for (const auto& element : elements(levelView)) {
+            BOOST_CHECK_EQUAL(element.index(), levelMapper.index(element));
+        }
+    }
+}
+
+void Opm::checkCellGlobalIdUniquenessForInteriorCells(const Dune::CpGrid& grid,
+                                                      const std::vector<std::shared_ptr<Dune::cpgrid::CpGridData>>& data)
+{
+    const auto& leafGridView = grid.leafGridView();
+    const int maxLevel = grid.maxLevel();
+
+    for (int level = 1; level <= maxLevel +1; ++level) {
+        bool isLeaf = (level == maxLevel+1);
+        const auto& levelOrLeafData = data[level];
+        const auto& levelOrLeafGlobalIdSet = levelOrLeafData->globalIdSet();
+        const auto& elements = isLeaf?
+            Dune::elements(leafGridView, Dune::Partitions::interior) :
+            Dune::elements(grid.levelGridView(level), Dune::Partitions::interior);
+
+        std::vector<int> levelOrLeaf_interior_cell_global_ids;
+        levelOrLeaf_interior_cell_global_ids.reserve(levelOrLeafData->size(0));
+
+        for (const auto& element : elements) {
+            levelOrLeaf_interior_cell_global_ids.push_back(levelOrLeafGlobalIdSet.id(element));
+        }
+
+        const int levelOrLeaf_local_interior_cell_count = levelOrLeaf_interior_cell_global_ids.size();
+        const int levelOrLeaf_global_cell_count = grid.comm().sum(levelOrLeaf_local_interior_cell_count);
+
+        // Gather all global IDs across MPI ranks.
+        const auto [all_levelOrLeaf_cell_ids, displLevelOrLeaf] = Opm::allGatherv(levelOrLeaf_interior_cell_global_ids, grid.comm());
+
+        const std::unordered_set<int> all_levelOrLeaf_cell_ids_set(all_levelOrLeaf_cell_ids.begin(), all_levelOrLeaf_cell_ids.end());
+
+        // Assert the expected size and uniqueness.
+        BOOST_CHECK_EQUAL(static_cast<int>(all_levelOrLeaf_cell_ids.size()), levelOrLeaf_global_cell_count);
+        BOOST_CHECK_EQUAL(all_levelOrLeaf_cell_ids.size(), all_levelOrLeaf_cell_ids_set.size());
+    }
+}
+
+void Opm::checkGridLocalAndGlobalIdConsistency(const Dune::CpGrid& grid,
+                                               const std::vector<std::shared_ptr<Dune::cpgrid::CpGridData>>& data)
+{
+    const auto& leafGridView = grid.leafGridView();
+    const auto& leafLocalIdSet = data.back()->localIdSet();
+
+    const bool isSerialRun = (grid.comm().size() == 0);
+
+    std::unordered_map<int, Dune::cpgrid::Entity<0>> leafElementLocalId;
+    std::unordered_map<int, Dune::cpgrid::Entity<3>> leafVertexLocalId;
```

Comment:
I think these should bet sets of ids. You are just checking for ids below.

Reply:
Mostly we check ids, but we check, for level grids, that the entities that do not vanish on the way do appear on the leaf:
 if (auto it = leafElementLocalId.find(localId); it != leafElementLocalId.end()) {
                    BOOST_CHECK(it->second.getLevelElem() == element);
                }
if (auto it = leafVertexLocalId.find(localId); it != leafVertexLocalId.end()) {
                    BOOST_CHECK(it->second.geometry().center() == vertex.geometry().center());
                }
So I would keep the maps with entities in this case.

Reply:
I missed that then